### PR TITLE
fix(auth,http): improve URI handling

### PR DESCRIPTION
**Pull request by savonarola**
Fixes [EMQX-12257](https://emqx.atlassian.net/browse/EMQX-12257)

Release version: v/e5.8.0

This PR addresses three entangled problems.

---

The first one is directly described in the task above. We have templated URIs in many parts of the code, and URI templates cannot be parsed with the standard URI parser (`uri_string`). So, multiple ad-hoc simplified parses are created, which may be error-prone. 

**We solve this by introducing a single loose parser** `emqx_utils_uri`.

---

The second one comes from the fact that the path that forms a part of the URI and the path that forms the "request-target" in the corresponding HTTP request are different concepts. The former may not start with a slash, while the latter MUST be absolute.

For example, in the valid URI `http://example.com?q=x`, the path is empty. However, the HTTP request `GET ?q=x HTTP/1.1` is invalid. It should be sent in the form `GET /?q=x HTTP/1.1`. See also: https://datatracker.ietf.org/doc/html/rfc7230#section-5.3

Browsers and high-level HTTP libraries automatically normalize the path, e.g.:
* https://github.com/sneako/finch/blob/71f469c46d3d8ea9993d986789cef8d14bcc819c/lib/finch/request.ex#L124
* https://github.com/benoitc/hackney/blob/7ae3831c1a56896d05d862de2c44cc751ae7790f/src/hackney_request.erl#L32

But `ehttpc` is a wrapper around quite a low-level HTTP client, `gun`, which does not make any URI parsing at all. And over `ehttpc` there are quite a lot of layers, so it was not quite clear which part of the code should be responsible for normalization. 

**We enforce normalization on** `emqx_bridge_http_connector` **level**.

---

Third, `emqx_bridge_http_connector` API is redundant. It accepts `base_url`, which is prepended to all requests, but:
* the `path` field of `base_url` is never used. Everywhere `base_url` is an explicitly cut-out scheme-host-port part.
* even if it were going to be used, the `path`-related part is not functional, because the correct validation and correct merging of paths is not implemented.

**So we do not accept** `base_url` **anymore and accept an** `request_base() :: #{scheme := term(), host := term(), port := term()}` **instead.**

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [x] Added property-based tests for code which performs user input validation (property-like ones)
- [ ] Changed lines covered in coverage report
- [x] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [x] Schema changes are backward compatible

[EMQX-12257]: https://emqx.atlassian.net/browse/EMQX-12257?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/emqx_auth/src/emqx_auth_utils.erl [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+%% API
+%%--------------------------------------------------------------------
+
+-spec parse_url(binary()) ->
+    {_Base :: emqx_utils_uri:request_base(), _Path :: binary(), _Query :: binary()}.
+parse_url(Url) ->
+    Parsed = emqx_utils_uri:parse(Url),
+    case Parsed of
+        #{scheme := undefined} ->
+            throw({invalid_url, {no_schema, Url}});
```
Comment (thalesmg) [inline apps/emqx_auth/src/emqx_auth_utils.erl:34]:
```suggestion
            throw({invalid_url, {no_scheme, Url}});
```

**Review by thalesmg — APPROVED** [review]

File: apps/emqx_auth/src/emqx_auth_utils.erl [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    Parsed = emqx_utils_uri:parse(Url),
+    case Parsed of
+        #{scheme := undefined} ->
+            throw({invalid_url, {no_scheme, Url}});
+        #{authority := undefined} ->
+            throw({invalid_url, {no_host, Url}});
+        #{authority := #{userinfo := Userinfo}} when Userinfo =/= undefined ->
+            throw({invalid_url, {userinfo_not_supported, Url}});
+        #{fragment := Fragment} when Fragment =/= undefined ->
+            throw({invalid_url, {fragments_not_supported, Url}});
```
Comment (zmstone) [inline apps/emqx_auth/src/emqx_auth_utils.erl:40]: why this is considered invalid ?

Comment (savonarola) [reply apps/emqx_auth/src/emqx_auth_utils.erl:40]: fragments specify locations in a fetched resource, but we do not use them anyhow

File: apps/emqx_utils/src/emqx_utils_uri.erl [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+%% > well-formed URI reference into its components.
+%%
+%% > ^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?
+%%
+%% We skip capturing some unused parts of the regex.
+
+-define(URI_REGEX,
+    ("^(?:(?<scheme>[^:/?#]+):)?(?<authority>//[^/?#]*)?"
+    "(?<path>[^?#]*)(?<query>\\?[^#]*)?(?<fragment>#.*)?")
+).
```
Comment (zmstone) [inline apps/emqx_utils/src/emqx_utils_uri.erl:215]: nit: move it to the top.

File: apps/emqx_utils/src/emqx_utils_uri.erl [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+
+uri_regexp() ->
+    {ok, RE} = re:compile(?URI_REGEX, [caseless]),
```
Comment (zmstone) [inline apps/emqx_utils/src/emqx_utils_uri.erl:218]: maybe worth caching it in persistent term in a on_load callback.
simple profiling shows re:compile is 20x as persistent_term:get

Comment (savonarola) [reply apps/emqx_utils/src/emqx_utils_uri.erl]: ✅ 
Currently, we do not parse URIs anywhere in hot loops only on resource/auth creation.
But let the code be a bit more foolproof and efficient.

File: apps/emqx_utils/src/emqx_utils_uri.erl (hunk 3 above)
Comment (savonarola) [reply apps/emqx_utils/src/emqx_utils_uri.erl]: ✅

**Review by zmstone — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-2506)

**Pull request merged**